File: dotnet/src/MyDotey.CircularBuffer/CounterBuffer.cs
using System;

namespace MyDotey.Util
{
    /**
     * @author koqizhao
     *
     * Jul 25, 2018
     */
    public class CounterBuffer<T> : TimeSequenceCircularBuffer<CounterBucket<T>>
    {
        public CounterBuffer(TimeSequenceCircularBufferConfig bufferConfig)
            : base(bufferConfig)
        {
        }

        protected override CounterBucket<T> NewBucket(long startTime, long ttl)
        {
            return new CounterBucket<T>(startTime, ttl);
        }

        public long Get(T identity)
        {
            long count = 0;
            ForEach(bucket => count += bucket.Get(identity));
            return count;
        }

        public void Increment(T identity)
        {
            CurrentBucket.Increment(identity);
        }

        public void Decrement(T identity)
        {
            CurrentBucket.Decrement(identity);
        }
    }
}

File: java/circular-buffer/src/main/java/org/mydotey/util/CounterBuffer.java
package org.mydotey.util;

import java.util.concurrent.atomic.AtomicLong;

/**
 * @author koqizhao
 *
 * Jul 25, 2018
 */
public class CounterBuffer<T> extends TimeSequenceCircularBuffer<CounterBucket<T>> {

    public CounterBuffer(TimeSequenceCircularBufferConfig bufferConfig) {
        super(bufferConfig);
    }

    @Override
    protected CounterBucket<T> createBucket(long startTime, long ttl) {
        return new CounterBucket<T>(startTime, ttl);
    }

    public long get(final T identity) {
        final AtomicLong count = new AtomicLong();
        forEach(bucket -> count.addAndGet(bucket.get(identity)));
        return count.get();
    }

    public void increment(T identity) {
        getCurrentBucket().increment(identity);
    }

    public void decrement(T identity) {
        getCurrentBucket().decrement(identity);
    }

}
